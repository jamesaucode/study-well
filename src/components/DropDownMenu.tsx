import React, { useState, useContext } from "react";
import styled from "styled-components";

interface SOptionProps {
  selected?: boolean;
}
interface OverlayProps {
  show?: boolean;
}

interface ContextTypes {
  selected: string;
  setSelected: React.Dispatch<React.SetStateAction<string>>;
  setIsExpanded: React.Dispatch<React.SetStateAction<boolean>>;
  isExpanded: boolean;
}

const DropDownContext = React.createContext<ContextTypes>(null);

const SOption = styled.span<SOptionProps>`
  background-color: ${({ selected }): string =>
    selected ? "#8610f9" : "#f5f5f5"};
  color: ${({ selected }): string => (selected ? "#fff" : "#333")};
  font-size: 0.8em;
  font-weight: 500;
  padding: 16px 24px;
  z-index: 100;
  &:hover {
    cursor: pointer;
    background-color: #8610f9;
    color: #fff;
  }
`;

const Wrapper = styled.div`
  width: 100%;
  max-width: 150px;
  position: relative;
  &:hover {
    cursor: pointer;
  }
`;
const MenuWrapper = styled.div`
  border-bottom-left-radius: 4px;
  border-bottom-right-radius: 4px;
  width: 100%;
  position: absolute;
  top: 36px;
  display: flex;
  flex-direction: column;
  z-index: 100;
`;
const Btn = styled.button`
  border: 0;
  padding: 10px 20px;
  margin-bottom: 1rem;
  font-size: 0.8em;
  font-weight: 600;
  text-align: left;
  z-index: 100;
`;
const Overlay = styled.div<OverlayProps>`
  position: fixed;
  height: 100vh;
  width: 100vw;
  top: 0;
  left: 0;
  right: 0;
  display: ${({ show }): string => (show ? "block" : "none")};
  z-index: 10;
`;

interface CompoundComponent<P> extends React.FunctionComponent<P> {
  Option: React.FunctionComponent<OptionPropTypes>;
}
interface MenuPropTypes {
  extraOnClick?: (event: React.MouseEvent<HTMLSpanElement>) => void;
}
interface OptionPropTypes {
  extraOnClick?: (event: React.MouseEvent<HTMLSpanElement>) => void;
}

const useDropDownContext = (): ContextTypes => {
  const context = useContext(DropDownContext);
  return context;
};

const DropDownMenu: CompoundComponent<MenuPropTypes> = (props): JSX.Element => {
  const [isExpanded, setIsExpanded] = useState(false);
  const [selected, setSelected] = useState("Choose");
  const value = {
    selected,
    setSelected,
    isExpanded,
    setIsExpanded,
  };

  const handleClick = (event: React.MouseEvent<HTMLSpanElement>): void => {
    setIsExpanded(!isExpanded);
    if (props.extraOnClick) {
      props.extraOnClick(event);
    }
  };
  return (
    <DropDownContext.Provider value={value}>
      <Overlay
        onClick={(): void => {
          setIsExpanded(false);
        }}
        show={isExpanded}
      />
      <Wrapper role="menu">
        <Btn onClick={handleClick}>{`${selected} â–¼`}</Btn>
        <MenuWrapper>{isExpanded ? props.children : null}</MenuWrapper>
      </Wrapper>
    </DropDownContext.Provider>
  );
};

const Option: React.FunctionComponent<OptionPropTypes> = ({
  children,
  extraOnClick = (): void => {},
}): JSX.Element => {
  const { selected, setSelected, setIsExpanded } = useDropDownContext();
  const handleClick = (event: React.MouseEvent<HTMLSpanElement>): void => {
    setSelected(children.toString());
    setIsExpanded(false);
    if (extraOnClick) {
      extraOnClick(event);
    }
  };
  return (
    <SOption
      selected={selected === children}
      role="button"
      onClick={handleClick}
    >
      {children}
    </SOption>
  );
};

DropDownMenu.Option = Option;

export default DropDownMenu;
