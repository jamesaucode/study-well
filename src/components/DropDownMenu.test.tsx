/* eslint-disable */
import { mount } from "enzyme";
import DropDownMenu from "./DropDownMenu";

describe("DropDownMenu", () => {
  it("Component renders", () => {
    const wrapper = mount(<DropDownMenu />);
    expect(wrapper).toMatchSnapshot();
  });
});
