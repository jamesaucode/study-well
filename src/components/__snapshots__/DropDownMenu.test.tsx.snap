// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DropDownMenu Component renders 1`] = `
<DropDownMenu>
  <DropDownMenu__Wrapper
    role="menu"
  >
    <StyledComponent
      forwardedComponent={
        Object {
          "$$typeof": Symbol(react.forward_ref),
          "attrs": Array [],
          "componentStyle": ComponentStyle {
            "componentId": "DropDownMenu__Wrapper-sc-1pyz6ng-1",
            "isStatic": true,
            "lastClassName": "irgNyQ",
            "rules": Array [
              "width:100%;max-width:250px;position:relative;",
            ],
          },
          "displayName": "DropDownMenu__Wrapper",
          "foldedComponentIds": Array [],
          "render": [Function],
          "styledComponentId": "DropDownMenu__Wrapper-sc-1pyz6ng-1",
          "target": "div",
          "toString": [Function],
          "warnTooManyClasses": [Function],
          "withComponent": [Function],
        }
      }
      forwardedRef={null}
      role="menu"
    >
      <div
        className="DropDownMenu__Wrapper-sc-1pyz6ng-1 irgNyQ"
        role="menu"
      >
        <DropDownMenu__Toggle
          onClick={[Function]}
        >
          <StyledComponent
            forwardedComponent={
              Object {
                "$$typeof": Symbol(react.forward_ref),
                "attrs": Array [],
                "componentStyle": ComponentStyle {
                  "componentId": "DropDownMenu__Toggle-sc-1pyz6ng-3",
                  "isStatic": true,
                  "lastClassName": "dUSfAA",
                  "rules": Array [
                    "border:0;border-radius:3px;color:",
                    "#222222",
                    ";padding:10px 20px;margin-bottom:1rem;font-size:0.8em;font-weight:600;font-family:",
                    "-apple-system, BlinkMacSystemFont, \\"Segoe UI\\", Roboto, Helvetica, Arial, sans-serif, \\"Apple Color Emoji\\", \\"Segoe UI Emoji\\", \\"Segoe UI Symbol\\"",
                    ";text-align:left;z-index:100;&:hover{cursor:pointer;}",
                  ],
                },
                "displayName": "DropDownMenu__Toggle",
                "foldedComponentIds": Array [],
                "render": [Function],
                "styledComponentId": "DropDownMenu__Toggle-sc-1pyz6ng-3",
                "target": "button",
                "toString": [Function],
                "warnTooManyClasses": [Function],
                "withComponent": [Function],
              }
            }
            forwardedRef={null}
            onClick={[Function]}
          >
            <button
              className="DropDownMenu__Toggle-sc-1pyz6ng-3 dUSfAA"
              onClick={[Function]}
            >
              Choose â–¼
            </button>
          </StyledComponent>
        </DropDownMenu__Toggle>
        <DropDownMenu__MenuWrapper
          expanded={false}
        >
          <StyledComponent
            expanded={false}
            forwardedComponent={
              Object {
                "$$typeof": Symbol(react.forward_ref),
                "attrs": Array [],
                "componentStyle": ComponentStyle {
                  "componentId": "DropDownMenu__MenuWrapper-sc-1pyz6ng-2",
                  "isStatic": false,
                  "lastClassName": "jaOGXc",
                  "rules": Array [
                    "border-bottom-left-radius:4px;border-bottom-right-radius:4px;background-color:#fff;box-shadow:0px 0px 20px rgba(0,0,0,0.15);display:flex;flex-direction:column;padding:",
                    [Function],
                    ";position:absolute;top:40px;width:100%;z-index:100;transition:transform 300ms ease;transform-origin:top left;transform:",
                    [Function],
                    ";input{border:none;border-radius:3px;background-color:#eee;font-size:",
                    "0.8em",
                    ";padding:10px;margin:10px 0;}",
                  ],
                },
                "displayName": "DropDownMenu__MenuWrapper",
                "foldedComponentIds": Array [],
                "render": [Function],
                "styledComponentId": "DropDownMenu__MenuWrapper-sc-1pyz6ng-2",
                "target": "ul",
                "toString": [Function],
                "warnTooManyClasses": [Function],
                "withComponent": [Function],
              }
            }
            forwardedRef={null}
          >
            <ul
              className="DropDownMenu__MenuWrapper-sc-1pyz6ng-2 jaOGXc"
            />
          </StyledComponent>
        </DropDownMenu__MenuWrapper>
      </div>
    </StyledComponent>
  </DropDownMenu__Wrapper>
</DropDownMenu>
`;
